### PR TITLE
Adding information on the umbraco startup process

### LINKED ISSUE #646: Umbraco Startup Process
Either cut'n'paste @Shazwazza's comment here or get him to expand on it. Absolute gold.

http://issues.umbraco.org/issue/U4-10967#comment=67-44622

### PR DESCRIPTION
**Pull request by dampee**
This fixes #646 

Moved text around to different file. I think we had too much "expert" information on different topic all related to the startup.  Adding more information about the bootmanager (as requested in #646) would make this a massive file. 

So I've added new files containing the new information:
*  understanding bootmanagers (some info from the youtrack issue added with some general knowledge, includes also the expert implementing IBootManager info from the original file)
* troubleshooting slow startup (from the youtrack issue)
* Extending UmbracoApplication which describes the alternative of implementing your own UmbracoApplication (global.asax)

### CODE REVIEW

File: Reference/Events/Application-Startup.md [hunk 1]
```diff
@@ -1,14 +1,14 @@
-# Application startup events & event registration
+# Adding startup events & event registration
 
 In order to bind to certain events in the Umbraco application you need to make these registrations during application startup. 
 
 ## Use ApplicationEventHandler to register events
 
 **Applies to: Umbraco 6.1.0+**
 
-_if you are using a version previous to 6.1.0, see [here](../Events/application-startup.md) for application event details_
```
Comment (dampee) [inline Reference/Events/Application-Startup.md]: This link references to itself - so i've removed this!

File: Reference/Events/Troubleshooting-Slow-Startup.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+* packages that do stuff on startup that don't need to 
+* custom code that do things on startup that don't need to
+
+_Troubleshooting tip: You can change your log4net config to *output Debug* level logs, this will give you detailed timing information on almost all aspects of startup so you can diagnose what might be running slow._
+
+# The request process
+
+Once booting is complete, then the request process occurs which is the [normal routing](../Routing/) phase.
+
+On first request the only thing that would be slower than the second request is resolving the URL to the published content item which is subsequently cached. I don't think this is a bottleneck though.
```
Comment (marcemarc) [inline Reference/Events/Troubleshooting-Slow-Startup.md]: do we need the comment about bottleneck though

File: Reference/Events/Troubleshooting-Slow-Startup.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+# The rendering process
+
+Once the request process is complete, then the controller for the route will be launched followed by the rendering of the razor views.
+
+If your rendering is slow, be sure to review which controller is used. By default the [RenderMvcController](../../Implementation/default-routing/Controller-Selection/) will be used.  But it is possible to implement [custom routing](../Implementation/Custom-Routing/) to have bad performing code in a custom controller.
```
Comment (marcemarc) [inline Reference/Events/Troubleshooting-Slow-Startup.md]: or indeed replace the default RenderMvcController at startup with a custom RenderMvcController that may contain bad performing code....

File: Reference/Events/Troubleshooting-Slow-Startup.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+There's several factors that can cause this part to be slow but this all comes down to implementation details. Especially if you have decided that caching will solve all of your issues. If that is the case, than the first request will always be slow and fixing that comes down to solving the performance issues that have been 'fixed' by putting caching in front of it.
```
Comment (marcemarc) [inline Reference/Events/Troubleshooting-Slow-Startup.md]: by putting caching in front of them.

File: Reference/Events/Understanding-Bootmanagers.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# Understanding Umbraco the BootManager
+
+After IIS started the W3 proces, umbraco will be launching.  There is a bootstrapper for the `Umbraco Application` which initializes all objects.
```
Comment (marcemarc) [inline Reference/Events/Understanding-Bootmanagers.md]: W3 process, Umbraco will begin launching

File: Reference/Events/Understanding-Bootmanagers.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+The responsible objects for the startup are the [CoreBootManager](../../apidocs/csharp/api/Umbraco.Core.CoreBootManager.html) and [WebBootManager](/apidocs/csharp/api/Umbraco.Web.WebBootManager.html) where the latter includes the Web portion of the application.
+
+The boot managers initialize the [UmbracoApplication](apidocs/csharp/api/Umbraco.Web.UmbracoApplication.html) (the global.asax) object.  After it has initialized the UmbracoApplication, it will initialize the ApplicationContext.  
+
+The bootmanager will initialize the ApplicationContext with a e.g. the database Context, services Context, profiling and logger. It will also register the Application Startup handlers which will execute later using the [ApplicationEventsResolver](https://our.umbraco.org/apidocs/csharp/api/Umbraco.Core.ObjectResolution.ApplicationEventsResolver.html).
```
Comment (marcemarc) [inline Reference/Events/Understanding-Bootmanagers.md]: with a e.g. ???

will initialize the ApplicationContext with: the database Context, services context...

File: Reference/Events/Understanding-Bootmanagers.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+For those wondering: **Examine actually doesn't do anything on startup if the indexes are built.**
```
Comment (marcemarc) [inline Reference/Events/Understanding-Bootmanagers.md]: Examine actually doesn't do anything on startup - 'if the indexes are built'

File: Reference/Events/Troubleshooting-Slow-Startup.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
```
Comment (marcemarc) [inline Reference/Events/Troubleshooting-Slow-Startup.md]: Troubleshooting Tip - Blank out all your templates and remove any custom startup code / packages - then reintroduce them incrementally to discover where the biggest impact on startup time is introduced.

File: Reference/Events/Troubleshooting-Slow-Startup.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+Once booting is complete, then the request process occurs which is the [normal routing](../Routing/) phase.
+
+On first request the only thing that would be slower than the second request is resolving the URL to the published content item which is subsequently cached. I don't think this is a bottleneck though.
+
+# The rendering process
+
+Once the request process is complete, then the controller for the route will be launched followed by the rendering of the razor views.
+
+If your rendering is slow, be sure to review which controller is used. By default the [RenderMvcController](../../Implementation/default-routing/Controller-Selection/) will be used.  But it is possible to implement [custom routing](../Implementation/Custom-Routing/) to have bad performing code in a custom controller.
+
```
Comment (marcemarc) [inline Reference/Events/Troubleshooting-Slow-Startup.md]: # Flexible Load Balancing

The umbracoCacheInstruction database table is checked for new publish and unpublish cache instructions during the startup process (or have I made this up?) - if you are running your environments with a shared database, (eg development and preview or staging and live) then instructions created during one environment maybe stored and ready to replay on the other environment. If you are experiencing a slow startup time, it is worth checking the contents of the umbracoServer and umbracoCacheInstruction database tables, to rule this niche situation out from being the cause.

File: Reference/Events/Troubleshooting-Slow-Startup.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
+# Troubleshooting slow startup
+
+Slow startup can occur if: 
+* no indexes exist
+* models builder is used and is using Live modes and it's the first time it's compiling (you can use DLL mode to avoid this entirely)
+* the umbraco cache file doesn't exist
```
[the remark this answers is not in the document]
Comment (dampee) [reply Reference/Events/Troubleshooting-Slow-Startup.md]: I'll add the file name as that is the default place. Would have added also a link, but can't find the right document.

File: Reference/Events/Troubleshooting-Slow-Startup.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+* the `/App_Data/TEMP/PluginCache` is empty or if one of your DLLs has changed. This means a new scan of plugins will occur. Plugin scanning has also been heavily optimized since v7.5 (IIRC)
+* packages that do stuff on startup that don't need to
```
[the remark this answers is not in the document]
Comment (dampee) [reply Reference/Events/Troubleshooting-Slow-Startup.md]: Copy-paste from Shannon explanation :-)  Will take yours.  👍

File: Reference/Events/Troubleshooting-Slow-Startup.md (hunk 2 above)
Comment (dampee) [reply Reference/Events/Troubleshooting-Slow-Startup.md]: Indeed.  Will remove it. 👍

File: Reference/Events/Troubleshooting-Slow-Startup.md [hunk 12]
```diff
@@ -0,0 +1,30 @@
+# Troubleshooting slow startup
+
+Slow startup can occur if: 
+* no indexes exist
+* models builder is used and is using Live modes and it's the first time it's compiling (you can use DLL mode to avoid this entirely)
+* the umbraco cache file doesn't exist
+* the `/App_Data/TEMP/PluginCache` is empty or if one of your DLLs has changed. This means a new scan of plugins will occur. Plugin scanning has also been heavily optimized since v7.5 (IIRC)
+* packages that do stuff on startup that don't need to 
+* custom code that do things on startup that don't need to
```
[the remark this answers is not in the document]
Comment (dampee) [reply Reference/Events/Troubleshooting-Slow-Startup.md]: 👍

File: Reference/Events/Troubleshooting-Slow-Startup.md (hunk 3 above)
Comment (dampee) [reply Reference/Events/Troubleshooting-Slow-Startup.md]: removed the word "indeed"

> Or the default RenderMvcController is replaced at startup with a custom RenderMvcController that may contain bad performing code.

File: Reference/Events/Troubleshooting-Slow-Startup.md (hunk 4 above)
Comment (dampee) [reply Reference/Events/Troubleshooting-Slow-Startup.md]: 👍

File: Reference/Events/Troubleshooting-Slow-Startup.md [hunk 13]
```diff
@@ -0,0 +1,30 @@
+# Troubleshooting slow startup
+
+Slow startup can occur if: 
+* no indexes exist
```
[the remark this answers is not in the document]
Comment (dampee) [reply Reference/Events/Troubleshooting-Slow-Startup.md]: 👍

File: Reference/Events/Understanding-Bootmanagers.md (hunk 6 above)
Comment (dampee) [reply Reference/Events/Understanding-Bootmanagers.md]: 👍

File: Reference/Events/Understanding-Bootmanagers.md (hunk 7 above)
Comment (dampee) [reply Reference/Events/Understanding-Bootmanagers.md]: 👍

**Comment by dampee** [comment]
@marcemarc Can you hit the merge button if you are fine with it?

File: Reference/Events/Application-Startup.md [hunk 14]
```diff
@@ -88,51 +88,18 @@ If you want more control over execution you can override these properties:
...
-            /// At the end of a handled request do something... 
-            /// </summary>            
-            void UmbracoApplication_PostRequestHandlerExecute(object sender, EventArgs e)
-            {
-                //Do something...
-            }
-        }
-    }
+## Related Links
+* [Troubleshooting Slow Startup](./Troubleshooting-Slow-Start)
```
Comment (Jeavon) [inline Reference/Events/Application-Startup.md]: this needs to be .md I think?

Comment (dampee) [reply Reference/Events/Application-Startup.md]: Uhm really?
Aren't the .md extentions messing up the links the deployed to OUR?  Or are they stripped automagically?

**Pull request merged**